src/Psl/DataStructure/QueueInterface.php

```diff
@@ -0,0 +1,55 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\DataStructure;
+
+use Psl;
+use Countable;
+
+/**
+ * An interface representing a queue data structure ( FIFO ).
+ *
+ * @template T
+ *
+ * @see https://en.wikipedia.org/wiki/FIFO_(computing_and_electronics)
+ */
+interface QueueInterface extends Countable
+{
+    /**
+     * Adds a node to the queue.
+     *
+     * @psalm-param T $node
+     */
+    public function enqueue($node): void;
+
+    /**
+     * Retrieves, but does not remove, the node at the head of this queue,
+     * or returns null if this queue is empty.
+     *
+     * @psalm-return null|T
+     */
+    public function peek();
+
+    /**
+     * Retrieves and removes the node at the head of this queue,
+     * or returns null if this queue is empty.
+     *
+     * @psalm-return null|T
+     */
+    public function pull();
```

Comment:
These can return `null|T` - but if you queue `null`, the queue won't really be empty.

Reply:
In that case `T` is nullable, assuming `T` is `null|string`, the return type is `null|(null|string)`.

so it is expected to return null if the head of the queue does contain null as a value.

Reply:
Yeah in that case the `dequeue()` method is better, but using that function is a bit more cumbersome since it throws exceptions.

Reply:
that's why they both exist. you can use whatever behavior you prefer, null or an exception.

personally i prefer `pull`, unless working with nullable values ( in which case i wouldn't know if the queue is empty or if its just a null value, unless i call count() )